### ISSUE TITLE
Doesn't work in production

### ISSUE THREAD

**Issue by Donald646** [Kakulukian/youtube-transcript#38]
This gets blocked in production

**Comment by emilthemaker** [comment]
Mind specifying what's happening?

**Comment by Donald646** [comment]
I'm using Next.js Supabase vercel

It seems like when I go to production, it's getting blocked. I get the error Transcript is Disabled, but I know that's not true, because it works locally.

I get the 

Error: An error occurred in the Server Components render. The specific message is omitted in production builds to avoid leaking sensitive details. A digest property is included in this error instance which may provide additional details about the nature of the error.

I am using server components for it. Even when I switch to API routes it still gets blocked.

**Comment by Donald646** [comment]
```import { NextResponse } from 'next/server';
import { YoutubeTranscript } from 'youtube-transcript';

export async function POST(request: Request) {
  const { videoUrl } = await request.json();

  try {
    const transcript = await YoutubeTranscript.fetchTranscript(videoUrl);
    let newTranscript = '';
    for (let i = 0; i < transcript.length; i++) {
      newTranscript += `Timestamp: ${transcript[i].offset}, Text: ${transcript[i].text}\n`;
    }
    return NextResponse.json({ transcript: newTranscript });
  } catch (error) {
    console.error('Error fetching transcript:', error);
    return NextResponse.json(
      { error: 'Failed to fetch video transcript. Please check the URL and try again.' },
      { status: 500 }
    );
  }
}
```

this is my code above

**Comment by Donald646** [comment]
I saw the other issue #11 and it seems like it's the issue I'm facing, but recommending it because there seems to be no good solution. If YouTube is blocking sites, then would this library be useless then?

**Comment by kellenmace** [comment]
I was encountering this issue on Vercel, too. I believe this issue exists because YouTube is returning different HTML for YouTube video pages depending on where the request is coming from. When I scrape a page like `https://www.youtube.com/watch?v=rB9ql0L0cUQ` on my local machine, the HTML page contains a script tag with a `ytInitialPlayerResponse` variable in it, and within that object is the URLs to the caption tracks (this is what the `youtube-transcript` library relies on to get the transcript URLs). When I scrape that same page in a Serverless Function on Vercel though, the `ytInitialPlayerResponse` object does not contain the caption track URLs, therefore the code that `youtube-transcript` runs errors out.

## Option 1
If all you need is the lines of text from the transcript, then the method described in #11 where you use the `youtubei.js` NPM package to access the transcript may work just fine.

## Option 2
If you need *all* of the transcript data, including the start time and duration for every line, the easiest solution is to run your `youtube-transcript` code somewhere else. You can create a Google Cloud Function that executes the code and call it from your app, for example. This isn't ideal, since you have to have this one bit of functionality live outside of your Vercel-hosted app, but it works.

**Comment by Donald646** [comment]
I see thank you. The durations are a pretty core part of my app. The commenter of the youtubei.js code said it was also possible to get the durations, so I will check that out first.

**Comment by terrytjw** [comment]
i am just experiencing this issue too in `prod`. works perfectly fine on `local`. @Donald646 did you manage to get it to work on `prod`?

**Comment by Donald646** [comment]
@terrytjw No I haven't, I don't think it works. I'm planning to switch to youtubei.js, but it seems like the issue is prevalent over there as well.

**Comment by terrytjw** [comment]
> I was encountering this issue on Vercel, too. I believe this issue exists because YouTube is returning different HTML for YouTube video pages depending on where the request is coming from. When I scrape a page like `https://www.youtube.com/watch?v=rB9ql0L0cUQ` on my local machine, the HTML page contains a script tag with a `ytInitialPlayerResponse` variable in it, and within that object is the URLs to the caption tracks (this is what the `youtube-transcript` library relies on to get the transcript URLs). When I scrape that same page in a Serverless Function on Vercel though, the `ytInitialPlayerResponse` object does not contain the caption track URLs, therefore the code that `youtube-transcript` runs errors out.
> 
> ## Option 1
> If all you need is the lines of text from the transcript, then the method described in #11 where you use the `youtubei.js` NPM package to access the transcript may work just fine.
> 
> ## Option 2
> If you need _all_ of the transcript data, including the start time and duration for every line, the easiest solution is to run your `youtube-transcript` code somewhere else. You can create a Google Cloud Function that executes the code and call it from your app, for example. This isn't ideal, since you have to have this one bit of functionality live outside of your Vercel-hosted app, but it works.

hey @kellenmace , other than Google Cloud Function, can aws lambda work too?

**Comment by terrytjw** [comment]
> @terrytjw No I haven't, I don't think it works. I'm planning to switch to youtubei.js, but it seems like the issue is prevalent over there as well.

@Donald646 have you tried the Google Cloud Function approach?

**Comment by Donald646** [comment]
> > @terrytjw No I haven't, I don't think it works. I'm planning to switch to youtubei.js, but it seems like the issue is prevalent over there as well.
> 
> @Donald646 have you tried the Google Cloud Function approach?

No I haven't yet, but from another issue I was looking at on Youtubei.js they said It didn't work either, but they weren't getting transcripts so I don't know

https://github.com/SuspiciousLookingOwl/youtubei/issues/113

**Comment by leandronorcio** [comment]
> hey @kellenmace , other than Google Cloud Function, can aws lambda work too?

I can confirm it’s not working on AWS Lambda.

**Comment by Donald646** [comment]
Yea this library is cooked, it's kinda useless if it doesn't work in production

**Comment by emilthemaker** [comment]
> > > @terrytjw No I haven't, I don't think it works. I'm planning to switch to youtubei.js, but it seems like the issue is prevalent over there as well.
> > 
> > 
> > @Donald646 have you tried the Google Cloud Function approach?
> 
> No I haven't yet, but from another issue I was looking at on Youtubei.js they said It didn't work either, but they weren't getting transcripts so I don't know
> 
> [SuspiciousLookingOwl/youtubei#113](https://github.com/SuspiciousLookingOwl/youtubei/issues/113)

Working for me with youtubei.js but no clue why 😅

I'm on Supabase Edge Functions. Transcript works, but some other properties appear broken.

**Comment by metaloozee** [comment]
Because YouTube is always changing how it operates and because it enforces sign-in, it is making it impossible for data scrapers to obtain the data, this is a global problem.

You can find a similar issue that simply returns the YouTube message "Sign in to confirm you are not a bot" if you check at a few different libraries, including ytdl and youtubei.js.

**Comment by thecoconutstudio** [comment]
It's working for me on a regular linux production server

**Comment by colouredFunk** [comment]
It's working for me on a regular linux production server

**Comment by metaloozee** [comment]
> It's working for me on a regular linux production server

wait till their system find out that you're just a bot

**Comment by emilthemaker** [comment]
I'm getting the bot login thing too but can still scrape transcripts

--------------
Emil Lienemann
--------------

founder

talktweak

***@***.***

emil.cx/meet ( https://emil.cx/meet )

**Comment by colouredFunk** [comment]
it's stopped working for me now too. I'm not seeing any error message, it just hangs....

**Comment by aleksa-codes** [comment]
Just noticed it does not work on Vercel production for me anymore as well... I get an error: "Transcript is disabled on this video". Locally it works fine.

**Comment by aleksa-codes** [comment]
> I saw the other issue #11 and it seems like it's the issue I'm facing, but recommending it because there seems to be no good solution. If YouTube is blocking sites, then would this library be useless then?

`youtubei.js` is working for me on production, I switched to it based on the code in that issue: https://github.com/Kakulukian/youtube-transcript/issues/11#issuecomment-2272608479. I managed to get the timestamps as well, if someone is interested I can share the code with the timestamps.

Edit: does not work for everything on production, I was not able to get `basicInfo` like: title, duration, description etc. but getting transcription still works. Again works locally and does not on Vercel prod in my case. Found out the hard way after I migrated all my code to use `youtubei.js` for all of the data I was getting in different ways. I get "Sign in required". Something probably related to: https://github.com/LuanRT/YouTube.js/issues/696

**Comment by Lipcsyy** [comment]
> > I saw the other issue #11 and it seems like it's the issue I'm facing, but recommending it because there seems to be no good solution. If YouTube is blocking sites, then would this library be useless then?
> 
> `youtubei.js` is working for me on production, I switched to it based on the code in that issue: [#11 (comment)](https://github.com/Kakulukian/youtube-transcript/issues/11#issuecomment-2272608479). I managed to get the timestamps as well, if someone is interested I can share the code with the timestamps.

It says video unavailable for me. Did this occur to you too?

**Comment by aleksa-codes** [comment]
> > > I saw the other issue #11 and it seems like it's the issue I'm facing, but recommending it because there seems to be no good solution. If YouTube is blocking sites, then would this library be useless then?
> > 
> > 
> > `youtubei.js` is working for me on production, I switched to it based on the code in that issue: [#11 (comment)](https://github.com/Kakulukian/youtube-transcript/issues/11#issuecomment-2272608479). I managed to get the timestamps as well, if someone is interested I can share the code with the timestamps.
> 
> It says video unavailable for me. Did this occur to you too?

I was getting that error only with: `enable_safety_mode: true` as option in `Innertube.create({})`. Remove it or set it to false, if that is the case.

**Comment by aleksa-codes** [comment]
> > > I saw the other issue #11 and it seems like it's the issue I'm facing, but recommending it because there seems to be no good solution. If YouTube is blocking sites, then would this library be useless then?
> > 
> > 
> > `youtubei.js` is working for me on production, I switched to it based on the code in that issue: [#11 (comment)](https://github.com/Kakulukian/youtube-transcript/issues/11#issuecomment-2272608479). I managed to get the timestamps as well, if someone is interested I can share the code with the timestamps.
> 
> It says video unavailable for me. Did this occur to you too?

@Lipcsyy Sorry to follow up, but I noticed that the code here https://github.com/Kakulukian/youtube-transcript/issues/11#issuecomment-2272608479 uses a URL in `youtube.getInfo()`. However, the [YouTube.js documentation](https://github.com/LuanRT/YouTube.js?tab=readme-ov-file#getinfotarget-client) doesn't mention using a URL as `target`. According to the docs, the function expects the video ID. 

Here's the function I'm using to extract the ID from the URL:

```ts
const getYouTubeVideoId = (input: string): string => {
  const regExp: RegExp =
    /(?:https?:\/\/)?(?:www\.)?(?:youtube\.com\/(?:.*[?&]v=|(?:v|e(?:mbed)?)\/|shorts\/|live\/)|youtu\.be\/)([a-zA-Z0-9_-]{11})/;
  const match: RegExpMatchArray | null = input.match(regExp);

  return match && match[1] ? match[1] : input;
};
```

There might be a simpler approach, but I hope this helps!

**Comment by adnjoo** [comment]
I have the same issue with Next.js deployed to Vercel

```
rn [Error]: [YoutubeTranscript] 🚨 Transcript is disabled on this video (Osj9tv8aOqM)
```

**Comment by tonymanh-dev** [comment]
Any solution for that?

**Comment by adnjoo** [comment]
@tonymanh-dev 

> Any solution for that?

https://www.npmjs.com/package/youtubei.js/

e.g. https://github.com/adnjoo/fast-youtube-summary/blob/main/web/app/(api)/summarize/route.ts#L3-L24

**Comment by tushar453030** [comment]
Hey @adnjoo 

I implemented the same logic as yours in my express server
![image](https://github.com/user-attachments/assets/9361358d-ecef-4ce1-9d91-6f14da5023bd)

but when hitting the APIs it gives error  InnertubeError: This video is unavailable.

What could be the reason?

**Comment by tushar453030** [comment]
@adnjoo 

By any change you ever came accros the below error

When hitting the APIs locally using postman or UI locally it works fine I gets the transcript but when hosted the backend on vercel or aws it gives error 

https://www.youtube.com/watch?v=EGkGRs6YhoM
YoutubeTranscriptDisabledError: [YoutubeTranscript] 🚨 Transcript is disabled on this video 

![image](https://github.com/user-attachments/assets/8aa9d6bf-af94-454c-8f3d-97d6a53e195b)

**Comment by mrgoonie** [comment]
Youtube has been updated, this script was cooked

**Comment by tushar453030** [comment]
Hii @mrgoonie,
Checkout this repo it has hosted site shared by @adnjoo. Its working the code is same as of mine.
https://github.com/adnjoo/fast-youtube-summary/tree/main

The only diff is his code is in next.js and mine on express.

**Comment by M-YasirGhaffar** [comment]
I think the error in production is due to the same IP address of the hosting server. I also saw this error in the logs of the server:

```
Error: YoutubeTranscriptTooManyRequestError: [YoutubeTranscript] 🚨 YouTube is receiving too many requests from this IP and now requires solving a captcha to continue
```

**Comment by mrgoonie** [comment]
> I think the error in production is due to the same IP address of the hosting server. I also saw this error in the logs of the server:
> 
> ```
> Error: YoutubeTranscriptTooManyRequestError: [YoutubeTranscript] 🚨 YouTube is receiving too many requests from this IP and now requires solving a captcha to continue
> ```

That's exactly the reason, as I mentioned earlier, Youtube has just updated and will block your server IP address if it crawls the Youtube URL too many times.

So I came up with a solution that adding a proxy layer in every fetch.

Here is the demo: https://app.digicord.site/youtube/transcript

There is a FREE API within the page in case someone need it. Cheers!

**Comment by mrgoonie** [comment]
> Hii @mrgoonie, Checkout this repo it has hosted site shared by @adnjoo. Its working the code is same as of mine. https://github.com/adnjoo/fast-youtube-summary/tree/main
> 
> The only diff is his code is in next.js and mine on express.

It's working because your IP address has not been blocked, not working for me because mine is cooked 😄

**Comment by swarajbachu** [comment]
so its not yet solved yet 🥲, I felt this was easy then saw the error on production

**Comment by ngocsangyem** [comment]
Hi @tushar453030, try to pass the video ID instead of the whole URL. It worked on my side by following this code.

```
const init = async () => {
    const youtube = await Innertube.create({
        lang: "en",
        location: "US",
        retrieve_player: false,
    });

    try {
        const info = await youtube.getInfo('axYAW7PuSIM');
        const transcriptData = await info.getTranscript();
        const mappedData = transcriptData.transcript.content.body.initial_segments.map(
            (segment) => segment.snippet.text
        );

        console.log('transcript', mappedData);

    } catch (error) {
        console.error("Error fetching transcript:", error);
        throw error;
    }
}
```

**Comment by hatemmezlini** [comment]
If anyone is still struggling with this. It's due to Youtube banning ISP Ips (why it would not work ion production) I have made a working solution using oauth2 that doesn't uses proxies and avoids Youtube ban. You can use it for free on Apify: https://apify.com/invideoiq/video-transcript-scraper. You only pay for Apify usage, however Apify gives you free credit of 5$ which will give you around 5000 transcripts

**Comment by swarajbachu** [comment]
> If anyone is still struggling with this. It's due to Youtube banning ISP Ips (why it would not work ion production) I have made a working solution using oauth2 that doesn't uses proxies and avoids Youtube ban. You can use it for free on Apify: https://apify.com/invideoiq/video-transcript-scraper. You only pay for Apify usage, however Apify gives you free credit of 5$ which will give you around 5000 transcripts

so you are directly using google apis ?

**Comment by hatemmezlini** [comment]
@swarajbachu No, There is an oauth plugin that was created by [yt-dlp](https://github.com/yt-dlp/yt-dlp/wiki/Extractors#logging-in-with-oauth) developers. It uses the Youtube on TV client because the token is never refreshed on TVs. All I had to do is a create a dummy account, it asked me for password once and than I saved the token and kept passing it in every request. I believe something similar can be developed here

**Comment by blake41** [comment]
> > I think the error in production is due to the same IP address of the hosting server. I also saw this error in the logs of the server:
> > ```
> > Error: YoutubeTranscriptTooManyRequestError: [YoutubeTranscript] 🚨 YouTube is receiving too many requests from this IP and now requires solving a captcha to continue
> > ```
> 
> That's exactly the reason, as I mentioned earlier, Youtube has just updated and will block your server IP address if it crawls the Youtube URL too many times.
> 
> So I came up with a solution that adding a proxy layer in every fetch.
> 
> Here is the demo: https://app.digicord.site/youtube/transcript
> 
> There is a FREE API within the page in case someone need it. Cheers!

what are you using for the proxy?

**Comment by mrgoonie** [comment]
There are plenty of them on the internet, I picked IPRoyal, good enough for me, but if you crawl too many, you will need many proxies to rotate 😅

**Comment by blake41** [comment]
> @swarajbachu No, There is an oauth plugin that was created by [yt-dlp](https://github.com/yt-dlp/yt-dlp/wiki/Extractors#logging-in-with-oauth) developers. It uses the Youtube on TV client because the token is never refreshed on TVs. All I had to do is a create a dummy account, it asked me for password once and than I saved the token and kept passing it in every request. I believe something similar can be developed here

I followed the extractor wiki and grabbed a proof of origin token. Is that the token you're referring to? I'm still getting blocked on my prod server even when I pass the PO using this library. Or which oauth plugin are you referring to with the yt-dlp library?

**Comment by petesampras12** [comment]
> > I think the error in production is due to the same IP address of the hosting server. I also saw this error in the logs of the server:
> > ```
> > Error: YoutubeTranscriptTooManyRequestError: [YoutubeTranscript] 🚨 YouTube is receiving too many requests from this IP and now requires solving a captcha to continue
> > ```
> 
> That's exactly the reason, as I mentioned earlier, Youtube has just updated and will block your server IP address if it crawls the Youtube URL too many times.
> 
> So I came up with a solution that adding a proxy layer in every fetch.
> 
> Here is the demo: https://app.digicord.site/youtube/transcript
> 
> There is a FREE API within the page in case someone need it. Cheers!

Fantastic! Is it possible to limit the request to only get the "content" part and not the "chunks"? Because it takes a long time to retreive the transcript for a 3 min video, but perhaps it would go quicker if i could only get "content"? Appreciate it!

**Comment by petesampras12** [comment]
Sadly both youtube-transcript and digicords variant has now stopped working (https://app.digicord.site/youtube/transcript).

Anyone else know of a working API?

**Comment by blake41** [comment]
i run all my requests through a residential proxy. unless you are doing an insane amount of requests, quite cheap

**Comment by savnani5** [comment]
> i run all my requests through a residential proxy. unless you are doing an insane amount of requests, quite cheap

I am trying this, but still getting blocked, what library are you using to fetch?

**Comment by blake41** [comment]
your requests must not be going through the proxy

import { setGlobalDispatcher, ProxyAgent } from 'undici';

// Load environment variables
dotenv.config();

// Set up global proxy if configured
if (process.env.PROXY_HOST && process.env.PROXY_PORT) {
  const proxyUrl = process.env.PROXY_USERNAME && process.env.PROXY_PASSWORD
    ? `http://${process.env.PROXY_USERNAME}:${process.env.PROXY_PASSWORD}@${process.env.PROXY_HOST}:${process.env.PROXY_PORT}`
    : `http://${process.env.PROXY_HOST}:${process.env.PROXY_PORT}`;
    
  console.log('Setting up global proxy dispatcher:', proxyUrl.replace(/:[^:@]*@/, ':***@'));
  const dispatcher = new ProxyAgent({ uri: proxyUrl });
  setGlobalDispatcher(dispatcher);
}

**Comment by petesampras12** [comment]
> > i run all my requests through a residential proxy. unless you are doing an insane amount of requests, quite cheap
> 
> I am trying this, but still getting blocked, what library are you using to fetch?

Did you get it to work?

**Comment by savnani5** [comment]
> your requests must not be going through the proxy
> 
> import { setGlobalDispatcher, ProxyAgent } from 'undici';
> 
> // Load environment variables dotenv.config();
> 
> // Set up global proxy if configured if (process.env.PROXY_HOST && process.env.PROXY_PORT) { const proxyUrl = process.env.PROXY_USERNAME && process.env.PROXY_PASSWORD ? `http://${process.env.PROXY_USERNAME}:${process.env.PROXY_PASSWORD}@${process.env.PROXY_HOST}:${process.env.PROXY_PORT}` : `http://${process.env.PROXY_HOST}:${process.env.PROXY_PORT}`;
> 
> console.log('Setting up global proxy dispatcher:', proxyUrl.replace(/:[^:@]_@/, ':_**@')); const dispatcher = new ProxyAgent({ uri: proxyUrl }); setGlobalDispatcher(dispatcher); }

I was using youtube-transcript and ytdl-core with proxies and it was throwing 500 internal serve error, got it working for now with innertube! Thanks

**Comment by savnani5** [comment]
> > > i run all my requests through a residential proxy. unless you are doing an insane amount of requests, quite cheap
> > 
> > 
> > I am trying this, but still getting blocked, what library are you using to fetch?
> 
> Did you get it to work?

Yes, working with youtubei.js (innertube)!

**Comment by petesampras12** [comment]
> > > > i run all my requests through a residential proxy. unless you are doing an insane amount of requests, quite cheap
> > > 
> > > 
> > > I am trying this, but still getting blocked, what library are you using to fetch?
> > 
> > 
> > Did you get it to work?
> 
> Yes, working with youutbei.js (innertube)!

Awesome! Is it this one? (https://github.com/haxzie/innerTube.js/)

I can't find any method that gets the transcript?

**Comment by savnani5** [comment]
> > > > > i run all my requests through a residential proxy. unless you are doing an insane amount of requests, quite cheap
> > > > 
> > > > 
> > > > I am trying this, but still getting blocked, what library are you using to fetch?
> > > 
> > > 
> > > Did you get it to work?
> > 
> > 
> > Yes, working with youutbei.js (innertube)!
> 
> Awesome! Is it this one? (https://github.com/haxzie/innerTube.js/)
> 
> I can't find any method that gets the transcript?

here this one: https://github.com/Kakulukian/youtube-transcript/issues/11#issuecomment-2272608479

**Comment by hatemmezlini** [comment]
> Sadly both youtube-transcript and digicords variant has now stopped working (https://app.digicord.site/youtube/transcript).
> 
> Anyone else know of a working API?

Don't need residential proxies with Innertube. We made a working API:
https://rapidapi.com/invideoiq-invideoiq-default/api/video-transcript-scraper

**Comment by mrgoonie** [comment]
> Sadly both youtube-transcript and digicords variant has now stopped working (https://app.digicord.site/youtube/transcript).
> 
> Anyone else know of a working API?

I ended up using this API service 😅
https://vidcap.xyz
(They also have API to capture screenshots of the video)

**Comment by mathiasfc** [comment]
> > @terrytjw No I haven't, I don't think it works. I'm planning to switch to youtubei.js, but it seems like the issue is prevalent over there as well.
> 
> @Donald646 have you tried the Google Cloud Function approach?

I tried here, sadly blocked too.

**Comment by jamesKhor** [comment]
have been working on this for a few hours. works locally, but not on vercel. i believe the issues was youtube is blocking vercel IP from scrapping. i will try to use #11  if it works. do it all over again. i only need the text

**Comment by mathiasfc** [comment]
@jamesKhor Just wishing you luck haha, and let us know if you manage to get the transcription working in production =)

**Comment by GowthamRaj24** [comment]
Working in dev but getting 500 error on Deployment, Can anyone share solution for this?

**Comment by myanpetra99** [comment]
> Working

did you use vercel ? i think youtube is banning vercel IP for some reason. deployed to another hosting works for me.

**Comment by MikePresman** [comment]
im getting rate limited on render.

**Comment by rafalzawadzki** [comment]
https://supadata.ai/youtube-transcript-api

**Comment by fetchworks-dev** [comment]
2026 state of this problem for anyone still landing here from search: "works locally, dies in production" is youtube bot-checking datacenter IPs at the player endpoint. your server (aws, gcp, render, railway, all of them) gets the "Sign in to confirm you're not a bot" interstitial instead of player data, so any library that parses captionTracks comes back empty. recent testing: roughly 9 in 10 videos fail from cloud IPs, 100% pass from residential egress. on top of that youtube started rolling out PO token requirements (exp=xpe caption urls) which no amount of proxying fixes.

DIY: rotating residential proxies (not datacenter) + budget for per-IP caption 429s around ~250-300/day. cookies don't help, it's IP reputation.

full disclosure, I maintain a hosted option that deals with all of this: https://apify.com/fetchworks/youtube-transcript-scraper, $2 per 1k transcripts, per-video honest statuses. writeup: https://github.com/Fetchworks/guides
